notebooks/spark-intro.ipynb

```diff
@@ -0,0 +1,1026 @@
+      "SLF4J: Found binding in [jar:file:/home/jovyan/.cache/coursier/v1/https/repo1.maven.org/maven2/org/slf4j/slf4j-log4j12/1.7.30/slf4j-log4j12-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]\n",
+      "SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.\n",
+      "SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]\n"
+     ]
+    },
+    {
+     "data": {
+      "text/html": [
+       "<a target=\"_blank\" href=\"http://2e4239bf6209:4040\">Spark UI</a>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "text/plain": [
+       "\u001b[32mimport \u001b[39m\u001b[36morg.apache.spark.sql._\u001b[39m\n",
+       "\u001b[36mspark\u001b[39m: \u001b[32mSparkSession\u001b[39m = org.apache.spark.sql.SparkSession@60d1ac27"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import org.apache.spark.sql._\n",
+    "\n",
+    "val spark = {\n",
+    "  NotebookSparkSession.builder()\n",
+    "    .master(\"local[4]\")\n",
+    "    .getOrCreate()\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "18596cc8-386d-4075-872f-722f6fd20930",
+   "metadata": {},
+   "source": [
+    "Because we're going to begin our introduction to Spark by using the older RDD interface, we're going to make available the `SparkContext` that is wrapped by `spark`.  We're going to declare this as an `implicit val` because some functions use an implicit argument to access it to keep us from needing to pass the value explicitly."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "f3b52a2d-8df1-4325-b4b7-21dc914aa29a",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "\u001b[36msc\u001b[39m: \u001b[32morg\u001b[39m.\u001b[32mapache\u001b[39m.\u001b[32mspark\u001b[39m.\u001b[32mSparkContext\u001b[39m = org.apache.spark.SparkContext@59b7590b"
+      ]
+     },
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "implicit val sc = spark.sparkContext"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "8de4369b-a5b6-4e96-a9bb-1e540778400f",
+   "metadata": {},
+   "source": [
+    "## Resilient Distributed Datasets (RDDs)\n",
+    "The heart and soul of Spark is the RDD.  This is an immutable, distributed data structure, where the data are spread across Spark's worker nodes (also called _executors_).  It is also fault tolerant (hence the R in RDD), in the fact that Spark maintains a call graph for each partition, so if an executor is lost, the partition can be shifted to another node and recalculated.\n",
```

Comment:
Probably good to introduce the acronym, just in case:
"The heart and soul of Spark is the Resilient Distributed Dataset (RDD).  This is an immutable, distributed data structure, where the data are spread across Spark's worker nodes (also called _executors_).  It is also fault tolerant (hence the R in RDD), in the fact that Spark maintains a call graph for each partition, so if an executor is lost, the partition can be shifted to another node and recalculated.\n",